### PR TITLE
Use ethcall and Multicall to load pool token balances

### LINKED ISSUE #391: Use ethers-multicall to batch contract calls
To batch calls like https://github.com/saddle-finance/saddle-frontend/blob/master/src/state/wallet/hooks.ts#L48

https://www.npmjs.com/package/ethers-multicall
https://github.com/makerdao/multicall
https://github.com/cavanmflynn/ethers-multicall

### PR DESCRIPTION
**Pull request by alphastorm**
https://github.com/Destiner/ethcall
https://github.com/makerdao/multicall

Closes #391

When running this locally make sure that your `.env.local` is set to:
```
REACT_APP_NETWORK_URL="http://127.0.0.1:8545/"
```

### CODE REVIEW

**Comment by penandlim** [comment]
Verified everything works on localhost and mainnet fork.

But I'm getting this error when accessing via fleek. Is this perhaps because fleek domain is not allowed in the alchemy API settings?

image

**Comment by alphastorm** [comment]
> Verified everything works on localhost and mainnet fork.
> 
> But I'm getting this error when accessing via fleek. Is this perhaps because fleek domain is not allowed in the alchemy API settings?
> 
> image

Yep, only saddle.exchange and saddlefinance.eth.link are allowed to use the Alchemy keys, if you want to test you an add the preview domain temporarily.

**Review by penandlim — APPROVED** [review]

File: src/state/wallet/hooks.ts [hunk 1]
```diff
@@ -1,98 +1,76 @@
...
-    async function pollBalance(): Promise<void> {
-      const newBalance = account
-        ? await tokenContract?.balanceOf(account)
-        : Zero
-      if (newBalance !== balance) {
-        setBalance(newBalance)
+    async function pollBalances(): Promise<void> {
+      if (!chainId) return
+
+      await ethcallProvider.init(getNetworkLibrary())
```
Comment (hammeiam) [inline src/state/wallet/hooks.ts:34]: I think you should be able to use `const {library} = useActiveWeb3React()` here instead

File: src/state/wallet/hooks.ts [hunk 2]
```diff
@@ -1,98 +1,76 @@
...
+        new Contract(WETH.addresses[chainId], ERC20_ABI),
+        new Contract(VETH2.addresses[chainId], ERC20_ABI),
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+      ].map((c): Call => c.balanceOf(account) as Call)
+      const balances: BigNumber[] = (await ethcallProvider.all(
+        balanceCalls,
+        {},
+      )) as BigNumber[]
+
+      setBalances({
```
Comment (hammeiam) [inline src/state/wallet/hooks.ts:58]: Could make token references less explicit with:

```
const tokens = Array.from(new Set(Object.keys(POOLS_MAP).map(({poolTokens}) => poolTokens).flat())) // get all unique tokens in active pools
const balanceCalls = tokens
  .map(t => new Contract(t.addresses[chainId], ERC20_ABI)) // instantiate multicall contract
  .map((c): Call => c.balanceOf(account) as Call); // make balance call
const balances = await ethcallProvider.all(balanceCalls, {})
setBalances(tokens.reduce((acc, t, i) => ({
  ...acc,
  [t.symbol]: balances[i]
}), {}))
```

File: src/state/wallet/hooks.ts [hunk 3]
```diff
@@ -1,98 +1,76 @@
...
-      void pollBalance()
+    if (account) {
+      void pollBalances()
     }
   }, BLOCK_TIME)
 
-  return balance
-}
-
-export function usePoolTokenBalances(
```
Comment (hammeiam) [inline src/state/wallet/hooks.ts:53]: kill it with fire

**Review by hammeiam — APPROVED** [review]
looks awesome! I'll also run this locally

**Pull request merged**